mlflow/runs.py

```diff
@@ -0,0 +1,74 @@
+"""
+CLI for runs
+"""
+import click
+import datetime
+import json
+from mlflow.entities import ViewType
+from mlflow.tracking import _get_store
+from tabulate import tabulate
+
+RUN_ID = click.argument("run_id", type=click.STRING)
+
+
+@click.group("runs")
+def commands():
+    """
+    Manage runs. To manage runs of experiments associated with a tracking server, set the
+    MLFLOW_TRACKING_URI environment variable to the URL of the desired server.
+    """
+    pass
+
+
+@commands.command("list")
+@click.option("--experiment_id", help="Specify the experiment for list of runs.", type=click.INT)
+@click.option("--view", "-v", default="active_only",
+              help="Select view type for list experiments. Valid view types are "
+                   "'active_only' (default), 'deleted_only', and 'all'.")
+def list_run(experiment_id, view):
+    """
+    List all runs of the specified experiment in the configured tracking server.
+    """
+    store = _get_store()
+    view_type = ViewType.from_string(view) if view else ViewType.ACTIVE_ONLY
+    runs = store.list_run_infos(experiment_id, view_type)
+    table = [[run.run_uuid, run.name, datetime.datetime.fromtimestamp(run.start_time / 1000.0).strftime('%Y-%m-%d %H:%M:%S')] for run in runs]
+    print(tabulate(sorted(table), headers=["Run Id", "Name", "Start time"]))
+
+
+@commands.command("delete")
+@RUN_ID
+def delete_run(run_id):
+    """
+    Mark a run for deletion. Return an error if the run does not exist or
+    is already marked. You can restore a marked run with ``restore_run``,
+    or permanently delete a run in the backend store.
+    """
+    store = _get_store()
+    store.delete_run(run_id)
+    print("Run with ID %s has been deleted." % str(run_id))
+
+@commands.command("restore")
+@RUN_ID
+def restore_run(run_id):
+    """
+    Restore a deleted run.
+    Returns an error if the run is active or has been permanently deleted.
+    """
+    store = _get_store()
+    store.restore_run(run_id)
+    print("Run with id %s has been restored." % str(run_id))
+
+@commands.command('export')
+@RUN_ID
+@click.argument('file', type=click.Path(exists=False))
+def export_run(run_id, file):
```

Comment:
This is cool! A few comments on the API:

- We probably shouldn't prescribe that users export the result to a file. A common use-case might be to just use some jq magic (e.g., `mlflow runs export --run-id abc123 | jq '.info.user_id'`), and users can always save to a file by using command line piping `mlflow runs export --run-id > output.json`). We can accomplish this by printing the data to stdout, and ensuring that all other output is printed to stderr (in mlflow we generally do this by using the logging library).
- `export` may be a bit of a misnomer. This API reminds me of [aws ec2 describe-instances](https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html), so maybe `describe` would be a reasonable name.
- In the [aws ec2 API](https://docs.aws.amazon.com/cli/latest/reference/ec2/index.html#cli-aws-ec2), they decided to make most operations which may apply to a single instance be able to apply to many instances. This probably makes sense for MLflow as well -- it's likely that if you want to export one run, maybe you want to export all of them, or some subset. There are a couple implications of this:
  - How do I provide multiple run ids? AWS uses `--instance-ids a b c`. This seems xargs-friendly (`cat file | xargs` would produce such a list) but I generally find it unintuitive that `--instance-ids` is plural when trying to look for a single instance (which is definitely a common use-case). I think that `mlflow runs describe --run-id a --run-id b` seems pretty obvious, although it may be a bit more cumbersome to construct programmatically.
  - What happens if I don't provide any run ids? Should we export all runs? This would require an experiment-id. I think this would be generally useful, but probably we could elide it for now, as it introduces interaction between providing experiment-id and run-ids.

If we go down the route of making this API apply to multiple runs, we may want to update `restore` and `delete` to do the same thing.

Reply:
I think the runs CLI implementation should be similar to the experiments CLI so in that case I would suggest that this API apply to just single run for now.